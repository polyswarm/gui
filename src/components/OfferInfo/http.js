class HttpOfferManage {
  constructor(url) {
    this.url = url;
  }

  pay(amount) {
    return new Promise((resolve) => {
      resolve(true);
    });
  }

  request(files)  {
    return new Promise((resolve) => {
      resolve(true);
    });
  }
}
export default HttpOfferManage;