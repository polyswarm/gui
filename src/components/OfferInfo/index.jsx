// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
// Bounty imports
import Header from '../Header';
import OfferPay from '../OfferPay';
import OfferMessageList from '../OfferMessageList';
import OfferRequest from '../OfferRequest';
import OfferSummary from '../OfferSummary';

class OfferInfo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      request: false,
      pay: false,
    };

    this.onBack = this.onBack.bind(this);
    this.onPayClick = this.onPayClick.bind(this);
    this.onRequestClick = this.onRequestClick.bind(this);
  }

  render() {
    const { props: { offer, address, addRequest, removeRequest, url, wallet,
      onBackPressed, requestsInProgress, onError },
    state: { request, pay } } = this;

    // only show actions if signing wallet is same as wallet used to create this
    let headerActions = [];
    if (address.toUpperCase() === offer.author.toUpperCase()) {
      headerActions = [
        {title: 'Pay', onClick: this.onPayClick},
        {title: 'Request', onClick: this.onRequestClick},
      ];
    }

    let last = '0';
    if (offer.messages) {
      const payments = offer.messages.filter(message => message.type === 'payment').sort((a, b) => a.amount < b.amount);
      last = payments.length > 0 ? payments[0].amount : '0';
    }

    return (
      <div className='OfferInfo'>
        {request && (
          <OfferRequest offer={offer}
            address={address}
            wallet={wallet}
            addRequest={addRequest}
            removeRequest={removeRequest}
            requestsInProgress={requestsInProgress}
            onBackPressed={this.onBack}
            onError={onError}
            onFilesSent={this.onBack}
            addMessage={this.onAddMessage}
            url={url}/>
        )}
        {pay && (
          <OfferPay
            offer={offer}
            last={last}
            address={address}
            wallet={wallet}
            addRequest={addRequest}
            removeRequest={removeRequest}
            requestsInProgress={requestsInProgress}
            onError={onError}
            onBackPressed={this.onBack}
            url={url}/>
        )}
        {!request && !pay && (
          <div className='Offer-Info'>
            <Header title={offer.guid}
              requests={requestsInProgress}
              back={true}
              onBack={onBackPressed}
              address={address}
              wallet={wallet}
              actions={headerActions}/>
            
            <div className='Offer-Info-Container'>
              <OfferSummary offer={offer}/>
              <OfferMessageList className='Offer-Info-Messages'
                offer={offer} />
            </div>
          </div>
        )}
      </div>
    );
  }

  addMessage() {
    const {props: {onAddMessage, offer}} = this;
    if (onAddMessage) {
      // Tells app that this offer has a new message to retrieve.
      onAddMessage(offer);
    }
  }

  onBack() {
    this.setState({request: false, pay: false});
  }

  onPayClick() {
    this.setState({pay: true});
  }

  onRequestClick() {
    this.setState({request: true});
  }
}

OfferInfo.propTypes = {
  offer: PropTypes.object.isRequired,
  wallet: PropTypes.object,
  address: PropTypes.string,
  onRequestWalletChange: PropTypes.func,
  addRequest: PropTypes.func,
  removeRequest: PropTypes.func,
  url: PropTypes.string,
  onBackPressed: PropTypes.func,
  requestsInProgress: PropTypes.array,
};
export default OfferInfo;
