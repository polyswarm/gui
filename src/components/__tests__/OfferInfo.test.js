import React from 'react';
import {render, mount} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import OfferInfo from '../OfferInfo';

const offer = {
  author: 'author',
  expert: 'expert',
  remaining: '100',
  closed: false,
  messages: [
    {
      guid: 'asdf',
      type: 'assertion',
      artifacts: [{name: 'Malicious', hash: 'asdf'}],
      verdicts: [true],
    },
    {
      guid: 'fdsa',
      type: 'assertion',
      artifacts: [{name: 'Benign', hash: 'fdsa'}],
      verdicts: [false],
    },
    {
      guid: 'blah',
      type: 'request',
      artifacts: [{name: 'Benign', hash: 'fdsa'}],
    },
    {
      guid: 'halb',
      type: 'request',
      artifacts: [{name: 'Malicious', hash: 'asdf'}],
    },
  ]
};

const walletList = [
  {address:'author', nct: '1', eth: '1'},
  {address:'demo', nct: '1', eth: '1'},
  {address:'omed', nct: '1', eth: '1'}
];

it('renders without crashing', () => {
  const wrapper = render(<OfferInfo 
    walletList={walletList}
    offer={offer}/>);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('shows OfferPay on pay click', () => {
  const wrapper = mount(<OfferInfo 
    walletList={walletList}
    offer={offer}/>);
  wrapper.find('.Header').find('.Button').first().simulate('click');

  expect(wrapper.find('.OfferPay')).toHaveLength(1);
});

it('shows OfferRequest on request click', () => {
  const wrapper = mount(<OfferInfo 
    walletList={walletList}
    offer={offer}/>);
  wrapper.find('.Header').find('.Button').last().simulate('click');

  expect(wrapper.find('.OfferRequest')).toHaveLength(1);
});

it('calls prop onAddMessage when addMessage is called', () => {
  const onAddMessage = jest.fn();
  const wrapper = mount(<OfferInfo
    walletList={walletList}
    offer={offer}
    onAddMessage={onAddMessage}
  />);
  const instance = wrapper.instance();

  instance.addMessage({});

  expect(onAddMessage).toHaveBeenCalledTimes(1);

});