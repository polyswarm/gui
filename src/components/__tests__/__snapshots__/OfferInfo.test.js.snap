// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="OfferInfo"
>
  <div
    class="Offer-Info"
  >
    <header
      class="Header"
    >
      <div
        class="Header-Row"
      >
        <div
          class="Header-Title"
        >
          <img
            alt="PolySwarm logo"
            class="Header-Logo"
            src="../public/img/back-arrow.svg"
          >
            
          </img>
          <h3>
            
          </h3>
          <span
            class="RequestSpinner"
          >
            
          </span>
        </div>
        <div
          class="Header-Actions"
        >
          <button
            class="Button header"
          >
            Pay
          </button>
          <button
            class="Button header"
          >
            Request
          </button>
        </div>
      </div>
      <div
        class="Header-Address"
      >
        <p
          class="Header-Divided"
          style="font-size:3.4285714285714284vw"
        >
          author
        </p>
        <p
          class="Header-Divided"
          style="font-size:3.4285714285714284vw"
        >
          Main: 1 NCT 1 ETH
        </p>
        <p
          style="font-size:3.4285714285714284vw"
        >
          Side: 1 NCT 
        </p>
      </div>
    </header>
    <div
      class="Offer-Info-Container"
    >
      <div
        class="OfferSummary"
      >
        <li
          class="StatRow-Vertical"
        >
          <p
            class="StatTitle"
          >
            Author:
          </p>
          <p
            class="StatContent"
          >
            author
          </p>
        </li>
        <li
          class="StatRow-Vertical"
        >
          <p
            class="StatTitle"
          >
            Expert:
          </p>
          <p
            class="StatContent"
          >
            expert
          </p>
        </li>
        <li
          class="StatRow-Vertical"
        >
          <p
            class="StatTitle"
          >
            Remaining Balance:
          </p>
          <p
            class="StatContent"
          >
            100 Nectar (NCT)
          </p>
        </li>
        <li
          class="StatRow-Vertical"
        >
          <p
            class="StatTitle"
          >
            
          </p>
          <p
            class="StatContent"
          >
            No
          </p>
        </li>
        <li
          class="StatRow-Vertical"
        >
          <p
            class="StatTitle"
          >
            Messages:
          </p>
          <p
            class="StatContent"
          >
            4
          </p>
        </li>
        <li
          class="StatRow-Vertical"
        >
          <p
            class="StatTitle"
          >
            Benign
          </p>
          <p
            class="StatContent"
          >
            Safe
          </p>
        </li>
        <li
          class="StatRow-Vertical"
        >
          <p
            class="StatTitle"
          >
            Malicious
          </p>
          <p
            class="StatContent"
          >
            Malicious
          </p>
        </li>
      </div>
      <div
        class="OfferMessageList"
      >
        <ul>
          <li>
            <div
              class="Card"
            >
              <header
                class="CardHeader Assertion-Malignant"
              >
                <div
                  class="CardHeader-Title"
                >
                  Assertion
                </div>
              </header>
              <div
                class="CardContent"
              >
                <ul>
                  <li
                    class="StatRow"
                  >
                    <p
                      class="StatTitle"
                    >
                      
                    </p>
                    <p
                      class="StatContent"
                    >
                      
                    </p>
                  </li>
                  <li
                    class="StatRow"
                  >
                    <p
                      class="StatTitle"
                    >
                      Malicious
                    </p>
                    <p
                      class="StatContent"
                    >
                      Malicious
                    </p>
                  </li>
                </ul>
              </div>
            </div>
          </li>
          <li>
            <div
              class="Card"
            >
              <header
                class="CardHeader Assertion-Benign"
              >
                <div
                  class="CardHeader-Title"
                >
                  Assertion
                </div>
              </header>
              <div
                class="CardContent"
              >
                <ul>
                  <li
                    class="StatRow"
                  >
                    <p
                      class="StatTitle"
                    >
                      
                    </p>
                    <p
                      class="StatContent"
                    >
                      
                    </p>
                  </li>
                  <li
                    class="StatRow"
                  >
                    <p
                      class="StatTitle"
                    >
                      Benign
                    </p>
                    <p
                      class="StatContent"
                    >
                      Safe
                    </p>
                  </li>
                </ul>
              </div>
            </div>
          </li>
          <li>
            <div
              class="Card"
            >
              <header
                class="CardHeader"
              >
                <div
                  class="CardHeader-Title"
                >
                  Request
                </div>
              </header>
              <div
                class="CardContent"
              >
                <ul>
                  <li
                    class="StatRow"
                  >
                    <p
                      class="StatTitle"
                    >
                      Files:
                    </p>
                    <p
                      class="StatContent"
                    >
                      Benign
                    </p>
                  </li>
                </ul>
              </div>
            </div>
          </li>
          <li>
            <div
              class="Card"
            >
              <header
                class="CardHeader"
              >
                <div
                  class="CardHeader-Title"
                >
                  Request
                </div>
              </header>
              <div
                class="CardContent"
              >
                <ul>
                  <li
                    class="StatRow"
                  >
                    <p
                      class="StatTitle"
                    >
                      Files:
                    </p>
                    <p
                      class="StatContent"
                    >
                      Malicious
                    </p>
                  </li>
                </ul>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
